internal/app/app.go

```diff
@@ -106,37 +100,55 @@ func NewDefaultRelayer(
 		proofSubmitter       = submit.NewSubmitterImpl(txSender, cfg.AllowKVCallbacks, neutronChain.PathEnd.ClientID)
 		txQuerier            = txquerier.NewTXQuerySrv(targetQuerier.Client)
 		trustedHeaderFetcher = trusted_headers.NewTrustedHeaderFetcher(neutronChain, targetChain, logger)
-		txProcessor          = txprocessor.NewTxProcessor(trustedHeaderFetcher, st, proofSubmitter, logger)
-		kvProcessor          = kvprocessor.NewKVProcessor(
+		txProcessor          = txprocessor.NewTxProcessor(
+			trustedHeaderFetcher, storage, proofSubmitter, logger, cfg.CheckSubmittedTxStatusDelay)
+		kvProcessor = kvprocessor.NewKVProcessor(
 			targetQuerier,
 			cfg.MinKvUpdatePeriod,
 			logger,
 			proofSubmitter,
-			st,
+			storage,
 			targetChain,
 			neutronChain,
 		)
-		txSubmitChecker = txsubmitchecker.NewTxSubmitChecker(
-			txProcessor.GetSubmitNotificationChannel(),
-			st,
-			neutronClient,
-			logger,
-			cfg.CheckSubmittedTxStatusDelay,
-		)
 		relayer = relay.NewRelayer(
 			cfg,
 			txQuerier,
-			st,
+			storage,
 			txProcessor,
 			kvProcessor,
-			txSubmitChecker,
 			logger,
 		)
 	)
 	return relayer
 }
 
-func loadChains(cfg config.NeutronQueryRelayerConfig, logger *zap.Logger) (neutronChain *cosmosrelayer.Chain, targetChain *cosmosrelayer.Chain, err error) {
+func NewDefaultStorage(cfg config.NeutronQueryRelayerConfig, logger *zap.Logger) (relay.Storage, error) {
+	var (
+		err            error
+		relayerStorage relay.Storage
+	)
+
+	if cfg.AllowTxQueries && cfg.StoragePath == "" {
+		logger.Fatal("RELAYER_DB_PATH must be set with RELAYER_ALLOW_TX_QUERIES=true")
+	}
+
+	if cfg.StoragePath != "" {
+		relayerStorage, err = storage.NewLevelDBStorage(cfg.StoragePath)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't initialize levelDB storage: %w", err)
+		}
+	} else {
+		relayerStorage = storage.NewDummyStorage()
```

Comment:
Do we even need dummy storage, maybe it is better to set it just to nil in case if we are not going to use it? That will also allow us to drop dummy storage implementation.

Reply:
> call storage.Close only when `cfg.AllowTxQueries==true`

done; i don't understand the "don't create" part. do you want me to pass `nil`s?

Reply:
yup, why not? this nil is not going to be referenced to because this means that we haven't allowed tx queries at all

Reply:
still scary af, don't like it : (

Reply:
also a tiny bit out of scope...

Reply:
well, that would simplify codebase since we wouldn't have to carry dummy storage around. since this is really out of scope, I am closing this thread.